packages/flutter/lib/src/material/app_bar.dart

```diff
@@ -1922,6 +2128,128 @@ class _RenderAppBarTitleBox extends RenderAligningShiftedBox {
   }
 }
 
+enum _ScrollUnderFlexibleVariant { medium, large }
+
+class _ScrollUnderFlexibleSpace extends StatefulWidget {
+  const _ScrollUnderFlexibleSpace({
+    this.title,
+    required this.variant,
+    this.centerCollapsedTitle,
+    this.primary = true,
+  });
+
+  final Widget? title;
+  final _ScrollUnderFlexibleVariant variant;
+  final bool? centerCollapsedTitle;
+  final bool primary;
+
+  @override
+  State<_ScrollUnderFlexibleSpace> createState() => _ScrollUnderFlexibleSpaceState();
+}
+
+class _ScrollUnderFlexibleSpaceState extends State<_ScrollUnderFlexibleSpace> {
+  @override
+  Widget build(BuildContext context) {
+    late final ThemeData theme = Theme.of(context);
+    final FlexibleSpaceBarSettings settings = context.dependOnInheritedWidgetOfExactType<FlexibleSpaceBarSettings>()!;
+    final double topPadding = widget.primary ? MediaQuery.of(context).viewPadding.top : 0;
+    final double collapsedHeight = settings.minExtent - topPadding;
+    final double scrollUnderHeight = settings.maxExtent - settings.minExtent;
+    final _ScrollUnderFlexibleConfig config;
+    switch (widget.variant) {
+      case _ScrollUnderFlexibleVariant.medium:
+        config = _MediumScrollUnderFlexibleConfig(context);
+        break;
+      case _ScrollUnderFlexibleVariant.large:
+        config = _LargeScrollUnderFlexibleConfig(context);
+        break;
+    }
+
+    late final Widget? collapsedTitle;
+    late final Widget? expandedTitle;
+    if (widget.title != null) {
+      collapsedTitle = config.collapsedTextStyle != null
+        ? DefaultTextStyle(
+            style: config.collapsedTextStyle!,
+            child: widget.title!,
+          )
+        : widget.title;
+      expandedTitle = config.expandedTextStyle != null
+        ? DefaultTextStyle(
+            style: config.expandedTextStyle!,
+            child: widget.title!,
+          )
+        : widget.title;
+    }
+
+    late final bool centerTitle;
+    {
```

Comment:
NICE. I like blocks.

Reply:
Yeah, I don't like polluting the namespace with functions that are only needed one place. In cases like this though I really wish Dart had some kind of 'switch expression' that could be used for cases where we just need a value based on an enum.